gordon/interfaces.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from zope.interface import Attribute
+from zope.interface import Interface
+
+
+class IEventMessage(Interface):
+    """A discrete unit of work for Gordon to process.
+
+    Gordon expects plugins to return or accept objects that implement this
+    interface in order to route them to other plugins, and handle retries
+    or cleanup in case of errors.
+    """
+
+    def __init__(msg_id, data, phase=None, history_log=None):
+        """Initialize an EventMessage.
+
+        Args:
+            msg_id (str): Unique message identifier.
+            data (dict): Data required to update DNS records.
+            phase (str): Current phase.
+            history_log (list): Log of actions performed on message.
+        """
+
+    def append_to_history(entry):
+        """Append entry to the IEventMessage's history_log.
+
+        Args:
+            entry (str): Information to append to log.
+        """
+
+
+class IGenericPlugin(Interface):
+    """**Do not** implement this interface directly.
+
+    Use :py:class:`gordon.interfaces.IEventConsumerClient`,
+    :py:class:`gordon.interfaces.IEnricherClient`,
+    or :py:class:`gordon.interfaces.IPublisherClient` instead.
+    """
+    phase = Attribute('Plugin phase')
```

Comment:
Just thinking out loud: I wish there was a way - either thru zope or elsewhere - to have an enum of supported values for `phase`. 🤷‍♀️

Reply:
I think this is possible through zope's [invariants](https://docs.zope.org/zope.interface/README.html#invariants) - it would be good if this invariant was the only place where these phases are defined (eg. `gordon.interfaces.PHASES`). Do you know if we will/already define them some place else, like the config?

Reply:
Constants might be a good idea. The invariant thing might be as well. I'd say let's move on and add/update if we find we need something like this.